File: 03-2.py
#!/usr/bin/env python

from functools import reduce
import re

input = open('03.input').read()

inst = re.compile('do\\(\\)|don\'t\\(\\)|mul\\([0-9]+,[0-9]+\\)')

res = 0
enabled = True

for i in re.findall(inst, input):
    match i:
        case 'do()':
            enabled = True
        case 'don\'t()':
            enabled = False
        case _:
            if enabled:
                res += reduce(lambda x, y: x * y,
                              map(lambda s: int(s),
                                  i[4:-1].split(',')))

print(res)

File: 03-2.rb
#!/usr/bin/env ruby

input = File.read('03.input')

inst = /do\(\)|don't\(\)|mul\([[:digit:]]+,[[:digit:]]+\)/

res = 0
enabled = true

input.scan(inst).each do |i|
  case i
  when 'do()'
    enabled = true
  when 'don\'t()'
    enabled = false
  else
    res += i[4..-2].split(',').map(&:to_i).reduce(&:*) if enabled
  end
end

puts res
